packages/Benchmarks/HybridTransport/src/DTK_Benchmark_CartesianMesh.hpp

```diff
@@ -0,0 +1,240 @@
+ * on a communicator of size 4. The mesh will be be fully replicated on each
+ * of the processors in the communicator (i.e. it will have no domain
+ * decomposition).
+ *
+ * Example 2: A mesh partitioned into 4 blocks and 1 set. This mesh will live
+ * on a communicator of size 4. The mesh will be subdivided onto each of the
+ * four processors in the communicator (i.e. it will have no replication).
+ *
+ * Example 3: A mesh partitioned into 2 blocks and 3 sets. This mesh will live
+ * on a communicator of size 6. The mesh will be replicated 3 times and each
+ * replica of the mesh will be subdivided onto 2 processors via domain
+ * decomposition.
+ *
+ * There are two ways to access the description of the mesh through this
+ * class:
+ *
+ * - The node coordinates and their global ids can be accessed - these nodes
+ *   will have ghosting.
+ *
+ * - The cells can be accessed via their global ids and either a connectivity
+ *   array or the coordinates of their centers - these cells will be uniquely
+ *   owned.
+ *
+ */
+class CartesianMesh
+{
+  public:
+    /*!
+     * \brief Default constructor.
+     */
+    CartesianMesh() { /* ... */}
+
+    /*
+     * \brief Data constructor.
+     *
+     * \param comm The parallel communicator over which the mesh is built.
+     *
+     * \param set_id The set id of the local mesh.
+     *
+     * \param block_id The block id of the local mesh.
+     *
+     * \param x_global_num_node The global number of nodes in the x direction.
+     *
+     * \param y_global_num_node The global number of nodes in the y direction.
+     *
+     * \param z_global_num_node The global number of nodes in the z direction.
+     *
+     * \param x_edge_offset The starting index of the nodes on this process in
+     * the x direction.
+     *
+     * \param y_edge_offset The starting indey of the nodes on this process in
+     * the y direction.
+     *
+     * \param z_edge_offset The starting indez of the nodes on this process in
+     * the z direction.
+     *
+     * \param local_x_edges The local edges of the mesh (node locations) in
+     * the x direction.
+     *
+     * \param local_y_edges The local edges of the mesh (node locations) in
+     * the y direction.
+     *
+     * \param local_z_edges The local edges of the mesh (node locations) in
+     * the z direction.
+     */
+    CartesianMesh( const Teuchos::RCP<const Teuchos::Comm<int>> &comm,
+                   const int set_id, const int block_id, const int num_i_blocks,
+                   const int num_j_blocks, const int num_k_blocks,
+                   const int x_global_num_node, const int y_global_num_node,
+                   const int x_edge_offset, const int y_edge_offset,
+                   const int z_edge_offset,
+                   const std::vector<double> &local_x_edges,
+                   const std::vector<double> &local_y_edges,
+                   const std::vector<double> &local_z_edges );
```

Comment:
It would be nice to find a way to clean this up - maybe embed some of the similar parameters in `std::array` objects? All of these are needed to index efficiently into the block-structured grid.